packages/frontend/src/app/dashboard.tsx

```diff
@@ -10,26 +10,67 @@ import LinearProgress, {
   linearProgressClasses,
 } from '@mui/material/LinearProgress';
 import { pdf } from '@react-pdf/renderer';
+import { v4 as uuidv4 } from 'uuid';
+import { BACKEND_STOP_JOB_URL } from './constants';
+import { styles } from './home.styles';
+import { Button } from '@mui/material';
 
 interface DashboardProps {
   onComplete: () => void;
   socket: WebSocket | null;
+  disconnectRef: React.MutableRefObject<boolean>;
   expectedItems: number;
 }
 
 // Each item from the websocket is an array of Metric objects.
-const Dashboard: React.FC<DashboardProps> = ({ onComplete, socket, expectedItems }) => {
+const Dashboard: React.FC<DashboardProps> = ({ onComplete, socket, disconnectRef, expectedItems }) => {
   // 'items' holds each item (which is an array of Metric objects) received from the socket.
   const [items, setItems] = useState<Metric[]>([]);
-  const [log, setLog] = useState<string>('');
+  const [log, setLog] = useState<string>('Log: Processing metrics...');
   const [error, setError] = useState<{ header: string; text: string }>({
     header: '',
     text: '',
   });
   const [showError, setShowError] = useState<boolean>(false);
+  const [retryButton, setRetryButton] = useState<JSX.Element | null>(null);
+
+
+  const buttonRetry = (
+    <button
+      onClick={() => {
+        window.location.reload();
```

Comment:
There's a react router specific way to do this use that instead

Reply:
Window.location is not really used, avoid using it as react router and so on normally have built in abstractions to handle it without breaking what they are doing